doc/developer-notes.md

```diff
@@ -115,6 +115,9 @@ code.
     Use `reinterpret_cast` and `const_cast` as appropriate.
   - Prefer [`list initialization ({})`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-list) where possible.
     For example `int x{0};` instead of `int x = 0;` or `int x(0);`
+  - Recursive functions are generally discouraged, and checked for by
```

Comment:
I think "generally" is too strong and also a bit of a contradiction with the next sentence. How about something like "Recursive functions are discouraged unless they are the best approach for a given problem. In this case  `NOLINTNEXTLINE(misc-no-recursion)` can be used to suppress the `clang-tidy` check."

Reply:
I do not think recursive functions should be "generally discouraged" or "discouraged". Consider this wording instead:

"Think twice before introducing recursive functions. Consider an iterative approach as well. When recursiveness is the best approach, use `NOLINTNEXTLINE(misc-no-recursion)` to suppress the check."

Reply:
> "Think twice before introducing recursive functions. Consider an iterative approach as well."

This just reads like a long form of "recursion is discouraged" to me.

Reply:
> I think "generally" is too strong

"generally" makes it less strong, since without it the sentence would imply it's always discouraged.

It's getting a bit too bikesheddy for me so I'm going to leave this conversation here. Imo all three suggestions so far pretty much achieve the same intent (and I'd be okay with): recursion should be used if it's clearly the best choice, otherwise don't.

Reply:
> "generally" makes it less strong, since without it the sentence would imply it's always discouraged.

I disagree and I did not suggest to just strike it but to use a different wording that is more clear. As it is now, I read this as "When you think you want to use recursion, you are generally wrong.". Maybe it's my ego but I don't like that statement ;) I suppose you read this as "When you want to iterate over anything, generally you shouldn't use recursion." and that is also a reasonable interpretation but I don't think the distinction is clear.